qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,213 @@
+                 input_state_circuit = None,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend]] = None):
+
+        """
+
+        Args:
+            num_evaluation_qubits: The number of qubits used in estimating the phase. The
+                                   phase will be estimated as a binary string with this many
+                                   bits.
+            pe_circuit: The phase estimation circuit.
+            num_unitary_qubits: Must agree with the number of qubits in the unitary in `pe_circuit` if
+                                `pe_circuit` is passed. This parameter will be set from
+                                `unitary` if `unitary` is passed.
+            unitary: The circuit representing the unitary operator whose eigenvalues (via phase) will
+                     be measured. Exactly one of `pe_circuit` and `unitary` must be passed.
+            input_state_circuit: The circuit that prepares the state whose eigenphase will be measured.
+                                 If this parameter is ommited, no preparation circuit will be run and
+                                 input state will be the all-zero state in the computational basis.
+            quantum_instance: The quantum instance on which the circuit will be run.
+        """
+
+        if unitary is None:
+            if pe_circuit is None:
+                raise ValueError('One one of `unitary` and `pe_circuit` may be `None`.')
+            else:
+                if num_unitary_qubits is None:
+                    raise ValueError('`num_unitary_qubits` is required when passing `pe_circuit`.')
+                self._pe_circuit = pe_circuit
+                self._num_unitary_qubits = num_unitary_qubits
+        else:
+            if not (num_unitary_qubits is None or num_unitary_qubits == unitary.num_qubits):
+                raise ValueError('`num_unitary_qubits` disagrees with size of `unitary`.')
+            pe_circuit = PhaseEstimation(num_evaluation_qubits, unitary)
+            self._num_unitary_qubits = unitary.num_qubits
+            self._pe_circuit = pe_circuit
+
+        self._num_evaluation_qubits = num_evaluation_qubits
+
+        if not input_state_circuit is None:
+            self._pe_circuit = pe_circuit.compose(
+                input_state_circuit,
+                qubits=range(num_evaluation_qubits, num_evaluation_qubits + self._num_unitary_qubits),
+                front=True)
+        else:
+            self._pe_circuit = pe_circuit
+
+        super().__init__(quantum_instance)
+        self._add_classical_register()
+
+    def _add_classical_register(self):
+        """Explicitly add measurement instructions only if we are using a state vector simulator."""
+        if not self._quantum_instance.is_statevector:
+            # Measure only the evaluation qubits.
+            regname = 'meas'
+            circ = self._pe_circuit
+            if not regname in [reg.name for reg in circ.cregs]:
```

Comment:
This is just a safeguard that this method is not called twice right? Otherwise we cannot know how people call their measurement registers, e.g. imagine the use case of somebody passing a PEC with measurements already attached. I'd rather remove this check completely because its not a safe way to check that measurements are already there.

Reply:
I had to put that in there. Otherwise, if you call `run` twice on the same instance (which you may want) an error is raised. Maybe this needs to be adjusted for the case that the user has already supplied the measurements.

Reply:
A cleaner solution would be to have a `construct_circuit` method which sets the a `self._circuit` attribute when called. Instead of re-constructing the circuit you then just return that cached circuit. 
Since all algorithms have a `construct_circuit` method we should probably add one here anyways (also if its quite trivial).

Reply:
Using `self._circuit`, which serves in part as a flag is a good solution.
But, I was mistaken when I said  "if you call run twice on the same instance". I actually added the check when I ran into the  the following. 1) A user instantiates a `PhaseEstimation` circuit `pec` (or the `PhaseEstimator` initialization does it), which has no measurement instructions. The user calls `PhaseEstimator` more than once , passing `pec` each time. Because `PhaseEstimator` does not make a copy, but rather mutates `pec`, an attempt may be made to add measurement instructions twice.
Checking if the register name exists is a not-great, but a so-so solution for ensuring that `PhaseEstimator` does not add measurement operators twice.  However, it doesn't work for checking that the user has not added measurements to `pec` before calling `PhaseEstimator`. 
I'm inclined to state in the docs that 1) the input pec may be mutated by adding measurements and 2) The user must not add measurements. The disadvantage is that the user may have some workflow with a source of circuits that have measurements already present. This may seem unlikely, but I wouldn't be surprised either.